src/coreclr/jit/fgopt.cpp

```diff
@@ -6018,8 +6026,9 @@ PhaseStatus Compiler::fgUpdateFlowGraphPhase()
 
     // Dominator and reachability sets are no longer valid.
     // The loop table is no longer valid.
-    fgDomsComputed    = false;
-    optLoopTableValid = false;
+    fgDomsComputed            = false;
+    optLoopTableValid         = false;
+    optLoopsRequirePreHeaders = false;
```

Comment:
I didn't quite follow the reset here. If we have `optLoopsRequirePreHeaders == true`, `fgUpdateFlowGraph()` will not touch the preheader block, but once that is done we reset it but for what reason? Is there any opportunity in future that where preheader can be optimized (removed/compacted) out or something?

Reply:
Probably they can get removed from "dead block elimination" that kicks in after lowering.

Reply:
Basically, I wanted it to be the case that if the loop table is invalid (`optLoopTableValid = false`) then `optLoopsRequirePreHeaders = false` also. There is no point to maintain the pre-headers if the loop table is invalid. In fact, after optimization, we want them to be optimized away, if they are empty.